### PR TITLE
Adds app verification alternative

### PR DESCRIPTION
**Pull request by protocol86**
Adds app verification alternative to verify phone number flow.

### CODE REVIEW

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 1]
```diff
@@ -76,21 +119,79 @@ - (void)verifyPhoneNumber:(NSString *)phoneNumber
...
+        dispatch_async(dispatch_get_main_queue(), ^{
+          completion(verificationID, error);
+        });
+      }
+    };
+    [self internalVerifyPhoneNumber:phoneNumber completion:^(NSString *_Nullable verificationID,
+                                                             NSError *_Nullable error) {
+      if (!error) {
+        // Associate the phone number with the verification ID.
+        verificationID.fir_authPhoneNumber = phoneNumber;
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Hasn't `fir_authPhoneNumber` already assigned in `verifyClientAndSendVerificationCodeToPhoneNumber:`?

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 2]
```diff
@@ -76,21 +119,79 @@ - (void)verifyPhoneNumber:(NSString *)phoneNumber
...
+        callBackOnMainThread(verificationID, nil);
         return;
       }
-      [self verifyClientAndSendVerificationCodeToPhoneNumber:phoneNumber
-                                 retryOnInvalidAppCredential:YES
-                                                    callback:callBackOnMainThread];
+      NSError *underlyingError = error.userInfo[NSUnderlyingErrorKey];
+      if (error.code != FIRAuthErrorCodeMissingAppToken &&
+          underlyingError.code != FIRAuthErrorCodeInvalidAppCredential) {
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Please check being internal error as well in the second case.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 3]
```diff
@@ -76,21 +119,79 @@ - (void)verifyPhoneNumber:(NSString *)phoneNumber
...
+      }
+      [self reCAPTCHAURLWithCompletion:^(NSURL *_Nullable reCAPTCHAURL,
+                                         NSError *_Nullable error) {
+        if (error) {
+          callBackOnMainThread(nil, error);
+          return;
+        }
+        FIRAuthURLCallbackMatcher callbackMatcher = ^BOOL(NSURL *_Nullable callbackURL) {
+          if (callbackURL) {
+            return [self isVerifyAppURL:callbackURL];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Please use `eventId` to start the flow and verify the same even ID here.

Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: If `isVerifyAppURL:` accepts nullable URL you can probably make the entire matcher block inline.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 4]
```diff
@@ -111,6 +212,90 @@ + (instancetype)providerWithAuth:(FIRAuth *)auth {
...
+  }
+  return NO;
+}
+
+/** @fn queryItemsForURLString
+    @brief Parses a URL into all available query items.
+    @param URLString The url to be parsed.
+    @return A dictionary of available query items in the target URL.
+ */
+- (NSDictionary<NSString *, NSString *> *)queryItemsForURLString:(NSString *)URLString {
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Use GTMNSDictionary+URLArguments instead?

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 5]
```diff
@@ -201,6 +386,59 @@ - (void)verifyClientWithCompletion:(FIRVerifyClientCallback)completion {
...
+  [self fetchAuthDomainWithCompletion:^(NSString *_Nullable authDomain,
+                                        NSError *_Nullable error) {
+    if (error) {
+      completion(nil, error);
+      return;
+    }
+    NSString *bundleID = [NSBundle mainBundle].bundleIdentifier;
+    NSString *clienID = _auth.app.options.clientID;
+    NSString *apiKey = _auth.requestConfiguration.APIKey;
+    NSString *languageCode = _auth.requestConfiguration.languageCode ?: @"en";
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Do not default to "en", which would override browser's default.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 6]
```diff
@@ -201,6 +386,59 @@ - (void)verifyClientWithCompletion:(FIRVerifyClientCallback)completion {
...
+  [FIRAuthBackend getProjectConfig:request
+                          callback:^(FIRGetProjectConfigResponse *_Nullable response,
+                                     NSError *_Nullable error) {
+    if (error) {
+      completion(nil, error);
+      return;
+    }
+    NSString *authDomain;
+    for (NSString *domain in response.authorizedDomains) {
+      NSInteger index = domain.length - kAuthDomainSuffix.length;
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Use `NSString's `hasSuffix:`.

File: Firebase/Auth/Source/FIRAuth.m [hunk 7]
```diff
@@ -986,8 +988,8 @@ - (BOOL)canHandleNotification:(NSDictionary *)userInfo {
 }
 #endif
 
-- (BOOL)canHandleURL:(NSURL *)url {
-  return NO;
+- (BOOL)canHandleURL:(NSURL *)URL {
+  return [_authURLPresenter canHandleURL:URL];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/FIRAuth.m]: Use dispatch_sync for public method calling private method synchronously.

File: Firebase/Auth/Source/FIRAuth.m [hunk 8]
```diff
@@ -387,6 +388,7 @@ - (instancetype)initWithApp:(FIRApp *)app {
       });
       return uid;
     };
+    _authURLPresenter = [[FIRAuthURLPresenter alloc] init];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/FIRAuth.m]: As discussed offline, consider using a shared FIRAuthURLPresenter so different Auth instances won't present over each other.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 9]
```diff
@@ -111,6 +212,90 @@ + (instancetype)providerWithAuth:(FIRAuth *)auth {
...
+- (BOOL)isVerifyAppURL:(NSURL *)URL {
+  NSArray *strings = [_auth.app.options.clientID componentsSeparatedByString:@"."];
+  NSString *reverseClientID =
+      [[strings reverseObjectEnumerator].allObjects componentsJoinedByString:@"."];
+  if (![[URL scheme] isEqualToString:reverseClientID]) {
+    return NO;
+  }
+  if (![[URL host] isEqualToString:kFirebaseAuthHost]) {
+    return NO;
+  }
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Please check path, too. And also port, user, and password to be cautious.

Alternatively, you can reconstruct the incoming URL without query and fragment by using NSURLComponent, and then compare to the expected URL directly.

File: Firebase/Auth/Source/FIRAuth.m (hunk 7 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: done, thanks!

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 6 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 5 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 10]
```diff
@@ -111,6 +212,90 @@ + (instancetype)providerWithAuth:(FIRAuth *)auth {
...
+    completion(nil, [FIRAuthErrorUtils missingPhoneNumberErrorWithMessage:nil]);
+    return;
+  }
+  [_auth.notificationManager checkNotificationForwardingWithCallback:
+      ^(BOOL isNotificationBeingForwarded) {
+    if (!isNotificationBeingForwarded) {
+      completion(nil, [FIRAuthErrorUtils notificationNotForwardedError]);
+      return;
+    }
+    FIRVerificationResultCallback callBackOnMainThread = ^(NSString *_Nullable verificationID,
```
[the remark this answers is not in the document]
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 4 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 9 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 3 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done.

Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: current tests does indicate that this is supported at all.  We can add it later after confirming with the web team.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 2 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 1 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Yes, removed.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 11]
```diff
@@ -44,6 +58,35 @@
...
+/** @var kFirebaseAuthHost
+    @brief The host part of the redirect URL after the app verification via reCAPTCHA.
+ */
+static NSString *const kFirebaseAuthHost = @"firebaseauth";
+
+/** @var kReCAPTCHAURLStringFormat
+    @brief The format of the URL used to open the reCAPTCHA page during app verification.
+ */
+NSString *const kReCAPTCHAURLStringFormat =
+    @"https://%@/__/auth/handler?apiKey=%@&authType=%@&ibi=%@&clientId=%@&hl=%@&v=iOS-Firebase-"
```
[the remark this answers is not in the document]
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done. Added the NSDictionary+URLArguments.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 12]
```diff
@@ -38,6 +38,8 @@
 #import "FIRSendVerificationCodeResponse.h"
 #import "FIRVerifyClientRequest.h"
 #import "FIRVerifyClientResponse.h"
+#import <GoogleToolboxForMac/GTMNSDictionary+URLArguments.h>
+#import <GTMSessionFetcher/GTMSessionFetcherService.h>
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: I don't see GTMSessionFetcher is used here?

File: Example/Auth/Tests/FIRApp+FIRAuthUnitTests.m [hunk 13]
```diff
@@ -28,7 +28,8 @@ + (FIROptions *)appOptions {
   return [[FIROptions alloc] initInternalWithOptionsDictionary:@{
     @"GOOGLE_APP_ID" : @"1:1085102361755:ios:f790a919483d5bdf",
     @"API_KEY" : @"FAKE_API_KEY",
-    @"GCM_SENDER_ID": @"217397612173"
+    @"GCM_SENDER_ID": @"217397612173",
+    @"CLIENT_ID" : @"123456.apps.googleusercontent.com"
```
Comment (XiangtianDai) [inline Example/Auth/Tests/FIRApp+FIRAuthUnitTests.m]: Please add a trailing comma after the last item so next time a new item is added after it the diff is going to be only one single line.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 1 above)
Comment (XiangtianDai) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Removing the assignment from `verifyClientAndSendVerificationCodeToPhoneNumber:...` affects the existing `verifyPhoneNumber:completion:` method. Why would we do that? It seems the assignment here should be removed instead.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 14]
```diff
@@ -111,6 +214,73 @@ + (instancetype)providerWithAuth:(FIRAuth *)auth {
...
+/** @fn isVerifyAppURL:
+    @brief Parses a URL into all available query items.
+    @param URL The url to be checked against the authType string.
+    @return Whether or not the URL matches authType.
+ */
+- (BOOL)isVerifyAppURL:(nullable NSURL *)URL {
+  if (!URL) {
+    return NO;
+  }
+  NSURLComponents *actualURLComponents = [NSURLComponents new];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]:
```
NSURLComponents *actualURLComponents = [NSURLComponents componentsWithURL:URL resolvingAgainstBaseURL:NO];
actualURLComponents.query = nil;
actualURLComponents.hash = nil;
```

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 15]
```diff
@@ -111,6 +214,73 @@ + (instancetype)providerWithAuth:(FIRAuth *)auth {
...
+  actualURLComponents.host = URL.host;
+  actualURLComponents.path = URL.path;
+  actualURLComponents.user = URL.user;
+  actualURLComponents.password = URL.password;
+
+  NSURLComponents *expectedURLComponents = [NSURLComponents new];
+  NSArray *strings = [_auth.app.options.clientID componentsSeparatedByString:@"."];
+  expectedURLComponents.scheme =
+      [[strings reverseObjectEnumerator].allObjects componentsJoinedByString:@"."];
+  expectedURLComponents.host = kFirebaseAuthURLHost;
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: I feel using the string literal inline here is actually more readable than using a constant, just like `@"deep_link_id"` below. Same for next line. But I don't feel strongly so it's up to you.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 16]
```diff
@@ -76,21 +124,76 @@ - (void)verifyPhoneNumber:(NSString *)phoneNumber
         });
       }
     };
+    [self internalVerifyPhoneNumber:phoneNumber completion:callBackOnMainThread];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Now you don't need the `callBackOnMainThread` variable but just using trailing block for better readability.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 17]
```diff
@@ -201,6 +371,66 @@ - (void)verifyClientWithCompletion:(FIRVerifyClientCallback)completion {
...
+    }
+    NSString *bundleID = [NSBundle mainBundle].bundleIdentifier;
+    NSString *clienID = _auth.app.options.clientID;
+    NSString *apiKey = _auth.requestConfiguration.APIKey;
+    NSMutableDictionary *urlArguments = [[NSMutableDictionary alloc] initWithDictionary: @{
+      @"apiKey" : apiKey,
+      @"authType" : kAuthTypeVerifyApp,
+      @"ibi" : bundleID,
+      @"clientId" : clienID,
+      @"v" : [NSString stringWithFormat:@"FirebaseAuth.iOS/%s %@",
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Please use a common method for this. Later when we're going to change it both need be changed.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 18]
```diff
@@ -201,6 +371,66 @@ - (void)verifyClientWithCompletion:(FIRVerifyClientCallback)completion {
...
+                                        FirebaseAuthVersionString,
+                                        GTMFetcherStandardUserAgentString(nil)]
+    }];
+    if (_auth.requestConfiguration.languageCode) {
+      urlArguments[@"hl"] = _auth.requestConfiguration.languageCode;
+    }
+    NSString *argumentsString = [urlArguments gtm_httpArgumentsString];
+
+    NSString *URLString =
+        [NSString stringWithFormat:kReCAPTCHAURLStringFormat,authDomain, argumentsString];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: space after first comma

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 6 above)
Comment (XiangtianDai) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: `if ([domain hasSuffix:kAuthDomainSuffix] && domain.length >= kAuthDomainSuffix.length + 2) {`

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 1 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: It was not removed from`verifyClientAndSendVerificationCodeToPhoneNumber:...` but rather from the end of this method, this was fixed.  This assignment however doesn't seem to be used at all anymore and will likely be removed in a follow up PR.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 6 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Great catch, thanks!

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 12 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Was used on line 391 to get part of the user agent string.
`@"v" : [NSString stringWithFormat:@"FirebaseAuth.iOS/%s %@",
                                        FirebaseAuthVersionString,
                                        GTMFetcherStandardUserAgentString(nil)]`

Replaced with  the class method on backend now, so removed.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 16 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Agreed, removed.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 14 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: done.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 15 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Agreed, these only make sense in this function for now.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 17 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Done, added as a class method to FIRAuthBackend.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m (hunk 18 above)
Comment (protocol86) [reply Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: done.

File: Example/Auth/Tests/FIRApp+FIRAuthUnitTests.m (hunk 13 above)
Comment (protocol86) [reply Example/Auth/Tests/FIRApp+FIRAuthUnitTests.m]: done.

**Pull request merged**